rust/src/rfb/rfb.rs

```diff
@@ -614,22 +650,97 @@ pub unsafe extern "C" fn rs_rfb_register_parser() {
+            0x2e, 0x6c, 0x6f, 0x63, 0x61, 0x6c, 0x64, 0x6f, 0x6d, 0x61, 0x69, 0x6e,
+        ];
+        let r = state.parse_request(buf);
+
+        assert_eq!(
+            r,
+            AppLayerResult {
+                status: -1,
+                consumed: 0,
+                needed: 0
+            }
+        );
+    }
+    #[test]
+    fn test_request_response_events() {
+        let mut init_state = RFBState::new();
+
+        let buf: &[u8] = &[
+            0x52, 0x46, 0x42, 0x20, 0x30, 0x30, 0x33, 0x2e, 0x30, 0x30, 0x38, 0x0a,
+            0x01, /* Number of security types: 1 */
+            0x02, /* Security type: VNC (2) */
+            0x02, /* Security type selected: VNC (2) */
+            0x54, 0x7b, 0x7a, 0x6f, 0x36, 0xa1, 0x54, 0xdb, 0x03, 0xa2, 0x57, 0x5c, 0x6f, 0x2a,
+            0x4e,
+            0xc5, /* 16 byte Authentication challenge: 547b7a6f36a154db03a2575c6f2a4ec5 */
+            0x00, 0x00, 0x00, 0x00, /* Authentication result: OK */
+            0x00, /* Share desktop flag: False */
+            0x05, 0x00, 0x03, 0x20, 0x20, 0x18, 0x00, 0x01, 0x00, 0xff, 0x00, 0xff, 0x00, 0xff,
+            0x10, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x1e, 0x61, 0x6e, 0x65, 0x61,
+            0x67, 0x6c, 0x65, 0x73, 0x40, 0x6c, 0x6f, 0x63, 0x61, 0x6c, 0x68, 0x6f, 0x73, 0x74,
+            0x2e, 0x6c, 0x6f, 0x63, 0x61, 0x6c, 0x64, 0x6f, 0x6d, 0x61, 0x69,
+            0x6e, /* Server framebuffer parameters */
+        ];
+        init_state.parse_response(&buf[0..12]);
+        let mut ok_state = parser::RFBGlobalState::TSClientProtocolVersion;
+        assert_eq!(init_state.state, ok_state);
+
+        init_state.parse_request(&buf[0..12]);
+        ok_state = parser::RFBGlobalState::TCSupportedSecurityTypes;
+        assert_eq!(init_state.state, ok_state);
+
+        init_state.parse_response(&buf[12..14]);
+        ok_state = parser::RFBGlobalState::TSSecurityTypeSelection;
+        assert_eq!(init_state.state, ok_state);
+
+        init_state.parse_request(&buf[14..15]);
+        ok_state = parser::RFBGlobalState::TCVncChallenge;
+        assert_eq!(init_state.state, ok_state);
+
+        init_state.parse_response(&buf[15..31]);
+        ok_state = parser::RFBGlobalState::TSVncResponse;
+        assert_eq!(init_state.state, ok_state);
+
+        init_state.parse_request(&buf[15..31]);
+        ok_state = parser::RFBGlobalState::TCSecurityResult;
+        assert_eq!(init_state.state, ok_state);
+
+        init_state.parse_response(&buf[31..35]);
+        ok_state = parser::RFBGlobalState::TSClientInit;
+        assert_eq!(init_state.state, ok_state);
+
+        init_state.parse_request(&buf[35..36]);
+        ok_state = parser::RFBGlobalState::TCServerInit;
+        assert_eq!(init_state.state, ok_state);
+
+        init_state.parse_response(&buf[36..90]);
+        ok_state = parser::RFBGlobalState::Message;
+        assert_eq!(init_state.state, ok_state);
```

Comment:
Nice work, Haleema! :)
Could you please tell the motivation behind a unit test testing the internal chained states of the parser?
It seems to me that the parser expects an input, runs it in a loop and moves from one state to another as a part of processing the same request.
Passing the parser an input from the middle of a request to validate a state looks like an invalid input to me. But, Idk anything about RFB so maybe it does make sense.. 🤷🏽‍♀️ and hence the question about your motivation. 😉

Reply:
I think I did this to try to test multiple states in both requests as I wasn't sure how the states are getting updated.
Further going into detail I found that the output of ` parse_request `will go as input to `parse_response` and so on.
The intention for testing the chained states was to test the flow from start to end. 
I was trying to think of it as 1 whole function instead of multiple states and was focusing on the flow.
Let me know what you think about this.

Also, I may still have some gaps in understanding but from what I have understood,when we are in `parse_request` function within the loop, once there is a match between the buffer state, the state gets updated and after that, there will be no match in the same function. The changed state will now have a match in `parse_response` function so the loop won't change the state again and the function just returns it.

Reply:
I _think_  that if your intention is the test the whole flow the best place for that would be Suricata-verify tests...

Reply:
So you are suggesting that I close this PR and write some s-v tests for these functions?

Reply:
For the purpose of what you were trying to do with this unittest here, yes. If you want to break down the test into smaller units, that is probably also fine and welcome, I would say. But let's try to keep those simpler :)